### ISSUE TITLE
Cototal copatterns

### ISSUE THREAD

**Issue by masaeedu** [Gabriel439/Haskell-Total-Library#12]
In Haskell we have patterns and pattern matching for deconstructing sum types. In some languages, there is also the dual concept of a "copattern" for building up product types.

Since your idea makes pattern matching "first class", and hence independent of the syntactic features of the language, the logical next step is to dualize the implementation and end up with [first class copatterns](https://github.com/masaeedu/cototal/blob/master/src/Cototal.hs#L28)!

Total patterns:

```haskell
module Total where

import Data.Void
import Data.Function ((&))

import Lens

class Empty a
  where
  impossible :: a -> x

instance Empty Void
  where
  impossible = absurd

instance (Empty a, Empty b) => Empty (Either a b)
  where
  impossible = either impossible impossible

_case :: Empty a => a -> x
_case = impossible

on :: Prism s t a Void -> (a -> o) -> (t -> o) -> s -> o
on (Prism _ m) l r = either r l . m

total :: Either Char Int -> String     -- Same as:
total = _case                          -- total = \case
  & on _Left  (\c -> replicate 3  c )  --   Left  c -> replicate 3 c
  & on _Right (\n -> replicate n '!')  --   Right n -> replicate n '!'
```

Cototal copatterns:

```haskell
module Cototal where

import Data.Function ((&))
import Control.Arrow ((&&&))
import Data.Maybe (listToMaybe)

import Lens

class Full a
  where
  trivial :: x -> a

instance Full ()
  where
  trivial = const ()

instance (Full a, Full b) => Full (a, b)
  where
  trivial = trivial &&& trivial

_cocase :: Full a => x -> a
_cocase = trivial

at :: Lens s t () b -> (i -> b) -> (i -> s) -> i -> t
at (Lens _ s) l r = s . (r &&& l)

cototal :: String -> (Maybe Char, Int)
cototal = _cocase
  & at _1 listToMaybe
  & at _2 length
```

**Comment by Gabriella439** [comment]
@masaeedu: This is pretty awesome!  I'd accept a pull request to add this

**Comment by Gabriella439** [comment]
Oh never mind, I see that you already have a separate package for this

**Comment by masaeedu** [comment]
@Gabriel439 No, that's just an experimental thing, definitely not something I'm planning to put on Hackage. I'd be happy to PR it to this repo.

**Comment by Gabriella439** [comment]
Alright, then I'd be happy to include it in this repository

**Comment by masaeedu** [comment]
Fixed by #13.
